rust/kernel/cpu.rs

```diff
@@ -0,0 +1,131 @@
+// SPDX-License-Identifier: GPL-2.0
+
+//! CPU utilities for SMP systems.
+//!
+//! Note: This module is a work in progress.
+
+use crate::{bindings, error::code, prelude::*};
+use core::{marker::PhantomData, mem::ManuallyDrop};
+
+/// Represents an SMP CPU by its zero-based kernel processor id.
+///
+/// Cpu values can be safely passed between threads (and across physical CPUs) as they are
+/// merely an integer index representing a CPU. All actual interaction with CPUs is late-
+/// bound.
+pub struct Cpu(i32);
+
+struct FnCall<F, R>
+where
+    F: FnOnce() -> R,
+{
+    out: Option<R>,
+    func: ManuallyDrop<F>,
+}
+
+unsafe extern "C" fn call_fn<F, R>(fn_box: *mut core::ffi::c_void)
+where
+    F: FnOnce() -> R,
+{
+    let fn_call = unsafe { &mut *(fn_box as *mut FnCall<F, R>) };
+    let func = unsafe { ManuallyDrop::take(&mut fn_call.func) };
+    fn_call.out = Some(func());
+}
+
+impl Cpu {
+    /// Constructs a new instance referring to a specific processor by id.
+    ///
+    /// # Safety
```

Comment:
Our convention is to use `# Safety` blocks in `unsafe` functions to document the pre-conditions that callers need to satisfy to use an unsafe function safely.

Given that this function is safe, you don't need a `# Safety` section. That is, the next paragraph can stand as is.

Reply:
Removed the `# Safety`, but kept the context as it seemed useful.